### ISSUE TITLE
Update Devfile Samples to reference v2.2.0 Parent stacks

### ISSUE THREAD

**Issue by kim-tsao** [devfile/api#1112]
## Which area this feature is related to?

 /area registry

## Which functionality do you think we should add?

The current samples have a parent reference to the default stack version in the registry.  e.g.  [Devfile sample go basic ](https://github.com/devfile-samples/devfile-sample-go-basic/blob/main/devfile.yaml#L18) points to the[ default go stack version ](https://registry.devfile.io/viewer/devfiles/community/go?devfile-version=1.0.2)in the community registry.   With this set up, parsing the samples will be successful because there are no conflicts.  If the sample were changed to point to the [latest version ](https://registry.devfile.io/viewer/devfiles/community/go?devfile-version=2.0.0)(v2.2.0 schema), parsing will fail because the parent and sample devfiles both contain the same command ids.  We should create a new version of the devfile sample that is compatible with the parent 2.2.0 schema.  

**Why is this needed? Is your feature request related to a problem?**

**Detailed description:**

**Comment by thepetk** [comment]
UPDATE:

@kim-tsao I'm not 100% about the workaround for this issue. In my mind I have two options:
1. Create a dir inside the sample repo and have different versions of devfiles (like we do in the registry)
2. Create a new branch that will host the new devfile which will point to the latest version of go stack)

**Comment by thepetk** [comment]
Resumed working on the issue.

**Target date**: 7 July 2023

**Comment by thepetk** [comment]
Working locally in order to check if the registry viewer supports sample versioning

**Comment by thepetk** [comment]
I've tried to check if the viewer supports multiple versions for samples:

* Created a new image locally for registry having the `nodejs` sample with two versions:
```yaml
versions:
  - version: 1.0.0
    schemaVersion: 2.0.0
    description: nodejs with devfile v2.0.0
    git:
      remotes:
        origin: https://github.com/redhat-developer/devfile-sample
  - version: 1.0.1
    schemaVersion: 2.2.1
    default: true
    description: nodejs with devfile v2.2.0
    git:
      remotes:
        origin: https://github.com/nodeshift-starters/devfile-sample
```

* I've built the image and deploy it on local minikube cluster but I wasn't able to see the 2 different versions of the sample in the UI.

**Comment by thepetk** [comment]
We have confirmed that the viewer does show up an updated image of the registry with multi-version samples.
Additionally it is confirmed that the index generator is able to handle multi-version samples. The investigation will continue on index server side (in order to be sure that the API is ok before started investigating on viewer side)

**Comment by thepetk** [comment]
After an investigation on the `registry-support` I have found that the multiple versions of a sample are supported. I think most probably we should investigate the viewer.

More detailed looking into `v2index/sample` I can see all updates from the image and both versions look ok.

**Comment by thepetk** [comment]
@michael-valdron I was investigating on the viewer side and found this https://github.com/devfile/devfile-web/blob/main/libs/core/src/functions/fetch-devfiles/fetch-devfiles.tsx#L38 which I think means that by default on the viewer we are not accepting samples having multiple versions. WDYT?

**Comment by michael-valdron** [comment]
> @michael-valdron I was investigating on the viewer side and found this https://github.com/devfile/devfile-web/blob/main/libs/core/src/functions/fetch-devfiles/fetch-devfiles.tsx#L38 which I think means that by default on the viewer we are not accepting samples having multiple versions. WDYT?

@thepetk Yeah that looks to be the culprit. I checked the `/v2index` again and I can see both versions for `code-with-quarkus` so it looks like a fix we need for the registry viewer.

**Comment by thepetk** [comment]
@michael-valdron do you think we should create a separate issue for the viewer or we should included it as part of this item?

**Comment by michael-valdron** [comment]
> @michael-valdron do you think we should create a separate issue for the viewer or we should included it as part of this item?

@thepetk I think this should be its own item, as a user story under the `area/registry-viewer` tag with a reference to this thread.

**Comment by thepetk** [comment]
Issue created here: https://github.com/devfile/api/issues/1184 , not sure if we should assign it the blocker label.

**Comment by thepetk** [comment]
I've tested different things while looking for the best workaround. The approach I'm most interested in is:

## Multiple sample versions inside extraDevfileEntries

As suggested above a good way to start is to include multiple sample versions inside the `extraDevfileEntries.yaml`. For example, for `go-basic` we can have two different versions, one pointing to the default `1.0.2` and one pointing to `2.1.0` which supports `schemaVersion >= 2.2.x`.

```yaml
- name: go-basic
  displayName: Basic Go
  description: Go 1.16 application
  icon: https://go.dev/blog/go-brand/Go-Logo/SVG/Go-Logo_Blue.svg
  tags:
    - Go
  projectType: Go
  language: Go
  provider: Red Hat
  versions:
    - version: 1.0.2
      schemaVersion: 2.1.0
      git:
        remotes:
          origin: https://github.com/devfile-samples/devfile-sample-go-basic.git
    - version: 2.1.0
      schemaVersion: 2.2.0
      default: true
      git:
        remotes:
          origin: https://github.com/thepetk/devfile-sample-go-basic.git
```

### Separate Branch, Separate Sample Repo OR Directory with devfiles?

If we follow the above format we will need to support different of devfile samples versions for `devfile.yaml`. I can see 3 approaches:

1. [Not Supported ATM] **Separate branch per devfile sample repo**: This way we can introduce a new branch for every sample repo (e.g. `2.2.x` default name) and upon `extraDevfileEntries.yaml` we can have:
```yaml
versions:
  - version: 2.1.0
    schemaVersion: 2.2.0
    git:
      checkoutFrom:
        revision: 2.2.x
      remotes:
        origin: https://github.com/devfile-samples/devfile-sample-go-basic.git
```
If I'm not missing something, right now we don't support this [in build-tools](https://github.com/devfile/registry-support/blob/main/build-tools/cache_samples.sh#L19-L31). If so, we will have to create a new issue to support this and then we can unblock this approach.

2. [Not Supported ATM] **Directory inside each samples that will introduce different devfile versions**: The idea is to have exactly the same structure as the stacks. So for every devfile sample we can have:
```
sample_root_dir/
...
---   devfiles/
       --- version1/
       --- version2/
       ...
       --- versionX/
```

3. **Use different repo per devfile.yaml version**: Following this solution we will have to use a different repo per devfile version. For example for the `go-basic`:
```yaml
  versions:
    - version: 1.0.2
      schemaVersion: 2.1.0
      git:
        remotes:
          origin: <repo1>
    - version: 2.1.0
      schemaVersion: 2.2.0
      default: true
      git:
        remotes:
          origin: <repo2>
```

**Comment by thepetk** [comment]
As discussed we will proceed with the option "Separate branch per devfile sample repo". As result, I've created an issue for the necessary actions we need to take for the registry-support #1209

**Comment by thepetk** [comment]
Moving to next sprint as the #1209 is currently under review

**Comment by thepetk** [comment]
I'll move this item to next sprint, as the blocker is still under review.

**Comment by thepetk** [comment]
Resuming working as the work needed to be done for https://github.com/devfile/api/issues/1209 is now completed

**Comment by thepetk** [comment]
As I'm working on the proposed workaround - [having multiple version branches inside a sample repo](https://github.com/devfile/api/issues/1112#issuecomment-1662160512) - I think we should create some rules for every new branch that represents a new sample version.

### Branch naming

The first thing we should consider is the naming of those branches. More detailed, I think they should follow a versioning format and every new version should be of type `.v{major}.{minor}.{bugfix}`

### Branch rules

Those branches should follow the same rules as the `main`/`master` (e.g the default) branch of the repo (more info [here](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule#about-branch-protection-rules)).

### Update Contributing Guide

As a result, we should also write those details down and update the registry contributing guide for this process in order to make everyone aware of this procedure.

**Comment by thepetk** [comment]
> As I'm working on the proposed workaround - [having multiple version branches inside a sample repo](https://github.com/devfile/api/issues/1112#issuecomment-1662160512) - I think we should create some rules for every new branch that represents a new sample version.
> 
> ### Branch naming
> The first thing we should consider is the naming of those branches. More detailed, I think they should follow a versioning format and every new version should be of type `v{major}.{minor}.{bugfix}`. Example branch: `v1.1.1`
> 
> ### Branch rules
> Those branches should follow the same rules as the `main`/`master` (e.g the default) branch of the repo (more info [here](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule#about-branch-protection-rules)).
> 
> ### Update Contributing Guide
> As a result, we should also write those details down and update the registry contributing guide for this process in order to make everyone aware of this procedure.

Atm the last actions to complete this issue are the ones mentioned here.

**Comment by thepetk** [comment]
Update for this item:

For every new sample version which will support the latest parent we will also update the sample port to be `8080` in all samples.

That said the following PRs have been created/updated:

* `python-basic`: https://github.com/devfile-samples/devfile-sample-python-basic/pull/35
* `java-springboot-basic`: https://github.com/devfile-samples/devfile-sample-java-springboot-basic/pull/33
* `code-with-quarkus`: https://github.com/devfile-samples/devfile-sample-code-with-quarkus/pull/23
* `dotnet-basic`: https://github.com/devfile-samples/devfile-sample-dotnet60-basic/pull/10
* `go-basic`: https://github.com/devfile-samples/devfile-sample-go-basic/pull/29/files

Two issues need to be addressed before creating those versions:

* **We already use port 8080 in `python-basic`, `java-springboot-basic` and `dotnet-basic`**: In order to fix that, I've created [PR#220](https://github.com/devfile/registry/pull/220).

* **We need to add protection to the version branches before merging**: As it is mentioned here https://github.com/devfile/api/issues/1112#issuecomment-1739306103 we will have to protect the following, already created branches:

- [ ] [devfile-sample-python-basic: v1.0.2](https://github.com/devfile-samples/devfile-sample-python-basic/tree/v1.0.2)
- [ ] [devfile-sample-java-springboot-basic: v1.2.2](https://github.com/devfile-samples/devfile-sample-java-springboot-basic/tree/v1.2.2)
- [ ] [devfile-sample-code-with-quarkus: v1.2.2](https://github.com/devfile-samples/devfile-sample-code-with-quarkus/tree/v1.2.2)
- [ ] [devfile-sample-dotnet60-basic: v1.1.2](https://github.com/devfile-samples/devfile-sample-dotnet60-basic/tree/v1.1.2)
- [ ] [devfile-sample-go-basic: v1.1.1](https://github.com/devfile-samples/devfile-sample-go-basic/tree/v1.1.1)

**Comment by thepetk** [comment]
I can confirm that all branches are protected, as only members of the devfile-samples org can merge PRs in any branch.

Other than that, in order to be able to use consistent port numbers I've created a PR (https://github.com/devfile/registry-support/pull/188) to update the library version, which will let us accept two components declaring the same port.

**Comment by thepetk** [comment]
As seen during testing, the ODC uses `index/sample` so it always fetches the `default` branch of each sample.

That said, in order to have always the latest version of every sample on the `default` branch we should first complete https://github.com/devfile/api/issues/1136. This update will allow us to have consistent ports (8080) on every latest sample version.

As a result, I'm putting it back to todo and putting the `blocked` label.

**Comment by thepetk** [comment]
The library version of the registry support has been updated and now we can support building a registry with a sample having multiple components using the same port

**Comment by thepetk** [comment]
Updating this item due to an issue found for parent overrides inside samples.

### New scope

This issue will focus on updating the `extraDevfileEntries.yaml` inside the registry by introducing a new sample version to every sample that has a parent with multiple stack versions. For example, the `go-basic` parent, `go` stack, has more than one versions so we will need to add a new `go-basic` version that will point to the latest parent version (atm 2.1.0).

### Default Version = Default Branch

For each sample we will keep the convention of maintaining the default sample version to the default sample branch. For example the default version of the go-basic sample (v1.0.2) will be inside the `main` branch (default branch of the repo).

### Latest Version = v2.x.x Branch

For every sample that needs to have a new version we will create a new branch starting with `v2.x.x.` format. This will be the latest version of the sample.

### Updated extraDevfileEntries.yaml

After the completion of the above steps the new `extraDevfileEntries.yaml` will look like:

```yaml
schemaVersion: 1.0.0
samples:
  - name: nodejs-basic
    displayName: Basic Node.js
    description: Node.js 16 application using Express 4.18.x
    icon: https://nodejs.org/static/images/logos/nodejs-new-pantone-black.svg
    tags:
      - Node.js
      - Express
    projectType: Node.js
    language: JavaScript
    provider: Red Hat
    versions:
      - version: 1.0.0
        default: true
        schemaVersion: 2.0.0
        git:
          revision: main           
          remotes:
            origin: https://github.com/redhat-developer/devfile-sample.git
  - name: go-basic
    displayName: Basic Go
    description: Go 1.16 application
    icon: https://go.dev/blog/go-brand/Go-Logo/SVG/Go-Logo_Blue.svg
    tags:
      - Go
    projectType: Go
    language: Go
    provider: Red Hat
    versions:
      - version: 1.1.0
        default: true
        schemaVersion: 2.2.0
        git:
          revision: main
          remotes:
            origin: https://github.com/devfile-samples/devfile-sample-go-basic.git
      - version: 2.0.0
        schemaVersion: 2.2.0
        git:
          revision: v2.0.0
          remotes:
            origin: https://github.com/devfile-samples/devfile-sample-go-basic.git
```

**Comment by thepetk** [comment]
Removing the `blocked` label as the updated scope is not affected anymore by the blocker.

Also I've created a new version for each sample having a parent stack with multiple versions:

* https://github.com/devfile-samples/devfile-sample-java-springboot-basic/pull/37
* https://github.com/devfile-samples/devfile-sample-python-basic/pull/35
* https://github.com/devfile-samples/devfile-sample-go-basic/pull/29

After those PRs are reviewed we can go ahead and create a PR for the updated `extraDevfileEntries.yaml`

**Comment by thepetk** [comment]
> Removing the `blocked` label as the updated scope is not affected anymore by the blocker.
> 
> Also I've created a new version for each sample having a parent stack with multiple versions:
> 
> * [Create springboot sample 2.0.0 version devfile-samples/devfile-sample-java-springboot-basic#37](https://github.com/devfile-samples/devfile-sample-java-springboot-basic/pull/37)
> * [Create version 2.0.0 for python sample devfile-samples/devfile-sample-python-basic#35](https://github.com/devfile-samples/devfile-sample-python-basic/pull/35)
> * [Create version 2.0.0 for go-basic sample devfile-samples/devfile-sample-go-basic#29](https://github.com/devfile-samples/devfile-sample-go-basic/pull/29)
> 
> After those PRs are reviewed we can go ahead and create a PR for the updated `extraDevfileEntries.yaml`

I'm moving this item to review status as the 3 sample related PRs are under review :)

**Comment by thepetk** [comment]
https://github.com/devfile/api/issues/1360 is now completed we can focus on the update of the `extraDevfileEntries.yaml`
